src/v/wasm/api.h

```diff
@@ -58,7 +58,7 @@ class factory {
     factory& operator=(const factory&) = delete;
     factory(factory&&) = delete;
     factory& operator=(factory&&) = delete;
-    virtual ss::future<std::unique_ptr<engine>> make_engine() = 0;
+    virtual ss::future<ss::shared_ptr<engine>> make_engine() = 0;
```

Comment:
General question not exactly related to this PR - what are the product implications of having a polymorphic `wasm::engine` interface? Is the idea for this to be configurable at some point to offer different features, performance characteristics, etc? Pointer to doc totally cool here.

Reply:
There are no product motivations for this currently.

The idea behind it is to be able to swap in different implementations (like I did in this PR) easily. It's unlikely that we'll want to do this dynamically at runtime, but the advantage is that no other code has to change if we change functionality like caching or multiplexing. So no code in the transform subsystem needed to change for this change to multiplexing. I suppose there are other ways of doing that without polymorphism, but it's really nice that the transform subsystem can mock out a [fake engine](https://github.com/redpanda-data/redpanda/blob/9a98aa7265945d4762ae617c376d166b4dd30733/src/v/transform/tests/test_fixture.h#L36-L49) that does nothing without the wasm needing to do anything.

Defining interfaces to subsystem boundaries like this allows for other subsystems to easily test or wrap behavior as needed in a decoupled manner.

There was at one point talk about letting users control which wasm runtime we use on a per transform basis - I think this is a bad idea for a few reasons, but it does allow the flexibility to do this if needed. However, this was not the motivation behind this - it was more the ability to having fake implementations for testing.

Reply:
Gotcha. I was wondering "why not `lw_shared`", but the rest of this PR makes it pretty clear. Besides which, I have to keep reminding myself that _neither_ of them is thread safe and the cost is just memory overhead.

Reply:
Ah, yes, from my understanding of `ss::lw_shared_ptr` doesn't work for what we're trying to do, we have to use `ss::shared_ptr`. Does that sound right to you?

(I also mix up `std::shared_ptr` vs `ss::shared_ptr` - glad it's not just me 😄 )

Reply:
Yeah that's my understanding. `ss::shared_ptr` has the virtual destructor so you pay the cost of whatever virtual dispatch machinery in exchange for runtime polymorphism.